### PR TITLE
RevGrid commit count: Do not reset when creating artificial commits

### PR DESCRIPTION
**Pull request by gerhardol**
## Proposed changes

The commit count handles the GitStatusMonitor that updates the commit count.
This is done in a number of situations, also when refreshing the grid.

RevGrid loading commits could be finished after the git status is finished
and should not reset the commit count.

This probably very seldom a problem in 3.5 or master before #9393, git-status is slightly delayed after the request, so the artificial commit that normally is added very early should be affected.
If you check out the first commit in a repo, it would be more common.
It occurs more often with #9393 that inserts the artificial commits after all other commits if HEAD is filtered out.
(Before #9393 the artificial commits were not displayed at all if HEAD was filtered, so the functionality did not exist.)

Note: Commit-count on the button is not affected by this, just the count in the grid. 
(There is a report in #9411 for something similar not affected.)
The count on the button must match the count in the grid, so the extra status init in the grid removed by this PR has no benefit.

Git-status could also be triggered when revision loading is completed, instead of starting after the short delay and running in parallel with git-log.

This is a candidate for 3.5.2 too.

### Before

image

### After

image

## Test methodology 

Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs [hunk 1]
```diff
@@ -2235,10 +2221,12 @@ void UpdateChangeCount(ObjectId objectId, IReadOnlyList<GitItemStatus> status)
             {
                 Debug.Assert(objectId == ObjectId.WorkTreeId || objectId == ObjectId.IndexId,
                     $"Unexpected Git object id {objectId}");
-                var changeCount = GetChangeCount(objectId);
                 var staged = objectId == ObjectId.WorkTreeId ? StagedStatus.WorkTree : StagedStatus.Index;
                 var items = status.Where(item => item.Staged == staged).ToList();
-                changeCount?.Update(items);
+
+                var changeCount = GetChangeCount(objectId);
+                Validates.NotNull(changeCount);
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:2228]: will fail on git error

**Review by mstv — COMMENTED** [review]
I have not grasped what triggers the update of the commit counts when the artificial have been added. How long can it be delayed?

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs (hunk 1 above)
Comment (gerhardol) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:2228]: No, just getting the object

```
        public ArtificialCommitChangeCount? GetChangeCount(ObjectId objectId)
        {
            return objectId == ObjectId.WorkTreeId
                ? _workTreeChangeCount
                : objectId == ObjectId.IndexId
                    ? _indexChangeCount
                    : null;
        }
```

**Comment by gerhardol** [comment]
> I have not grasped what triggers the update of the commit counts when the artificial have been added. How long can it be delayed?

Much to complicated to explain here, but status monitor is stopped prior to opening a new repo and opening Settings, Commit.
Started after opening the repo (and a few other commands that may affect the commit count like reset), then an interactive request is submitted, waiting 200-300 ms (to not interfere so much with other commands, yet be hardly noticable). 
When the filesystem is updated outside GE, the delay is 1s, subsequent changes 30s.

**Comment by mstv** [comment]
> > I have not grasped what triggers the update of the commit counts when the artificial have been added. How long can it be delayed?

> Much to complicated to explain here, but status monitor is stopped prior to opening a new repo and opening Settings, Commit.
Started after opening the repo (and a few other commands that may affect the commit count like reset), then an interactive request is submitted, waiting 200-300 ms (to not interfere so much with other commands, yet be hardly noticable).
When the filesystem is updated outside GE, the delay is 1s, subsequent changes 30s.

Thank you.
This PR seems to be necessary. We will see how long it delays the updates. Can be improved later if necessary.

**Comment by mstv** [comment]
The MC needs to be resolved.

**Comment by gerhardol** [comment]
> The MC needs to be resolved.

Was working on that...
Also made the `ArtificialCommitChangeCount` objects `readonly`

> This PR seems to be necessary. We will see how long it delays the updates. Can be improved later if necessary.

It is necessary to present the improvements in #9393 (before this was probably not an issue).

Hmm, the commit count button is not gray when it should be just after pushing `F5`...

**Review by mstv — APPROVED** [review]

**Comment by gerhardol** [comment]
Plan to squash merge tomorrow, also for 3.5

**Pull request merged**